controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +279,134 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachmentForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		engine, err := vbc.getVolumeEngine(vol)
+		if err != nil {
+			return err
+		}
+		if engine == nil {
+			vbc.logger.Warnf("Volume %v engine not found, skip offline rebuilding", volName)
+			delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+			return nil
+		}
+		if engine.Status.ReplicaModeMap == nil {
+			// wait for engine status synced
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol, engine)
+		if err != nil {
+			return err
+		}
+		if isVolumeInRebuilding {
+			return nil
+		}
+	}
 
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuilding longhorn.VolumeOfflineRebuilding) (bool, error) {
+	if offlineRebuilding == longhorn.VolumeOfflineRebuildingEnabled {
+		return true, nil
+	}
+
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return offlineReplicaRebuildingValue && offlineRebuilding != longhorn.VolumeOfflineRebuildingDisabled, nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachmentForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	areAllReplicasHealthy, err := vbc.isVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas)
+	if err != nil {
+		return va, err
+	}
+	if !areAllReplicasHealthy {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.AnyValue, longhorn.AttacherTypeVolumeRebuildingController)
+	}
+	return va, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas int, replicas map[string]*longhorn.Replica) (bool, error) {
+	if len(replicas) < numberOfReplicas {
+		return false, nil
+	}
+
+	healthyCount := 0
+	for _, replica := range replicas {
+		if replica.Spec.NodeID != replica.Status.OwnerID {
+			continue
+		}
+		if replica.Spec.FailedAt != "" {
+			continue
+		}
+		// the volume might be newly created and not attached yet.
+		if replica.Spec.HealthyAt != "" || (replica.Spec.LastFailedAt == "" && replica.Spec.LastHealthyAt == "") {
+			healthyCount++
+		}
+	}
+	return healthyCount >= numberOfReplicas, nil
+}
+
+func (vbc *VolumeRebuildingController) getVolumeEngine(vol *longhorn.Volume) (*longhorn.Engine, error) {
+	engines, err := vbc.ds.ListVolumeEngines(vol.Name)
+	if err != nil {
+		return nil, err
+	}
+	engine, err := vbc.ds.PickVolumeCurrentEngine(vol, engines)
+	if err != nil {
+		return nil, err
```

Comment:
Can the error be `not found`?

Reply:
No, it picks the engine from `engines`, and if it is unable to find the engine, it returns a customized error, https://github.com/longhorn/longhorn-manager/blob/c313e6d7666f58a09222075b906e34348938b4cb/datastore/longhorn.go#L1285-L1287